OpenSim/Examples/ReadXsensData/readXsensData.cpp

```diff
@@ -0,0 +1,94 @@
+/* -------------------------------------------------------------------------- *
+ *                            OpenSim:  exampleXsensDataReader.cpp            *
+ * -------------------------------------------------------------------------- *
+ * The OpenSim API is a toolkit for musculoskeletal modeling and simulation.  *
+ * See http://opensim.stanford.edu and the NOTICE file for more information.  *
+ * OpenSim is developed at Stanford University and supported by the US        *
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2019 Stanford University and the Authors                *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include "OpenSim/Common/DataAdapter.h"
+#include "OpenSim/Common/XsensDataReader.h"
+#include "OpenSim/Common/STOFileAdapter.h"
+
+using namespace std;
+
+static void PrintUsage(const char *aProgName, ostream &aOStream);
+
+int main(int argc, char* argv[]) {
+    using namespace OpenSim;
+    if (argc < 2) {
+        PrintUsage(argv[0], cout);
+        return(-1);
+    }
+    std::string folder = "";
+    std::string trial = ""; 
+    std::string xsensReaderConfigFile = "";
+    for (int i = 1; i < argc; i++) {
+        string option = argv[i];
+        if (option == "-F")
+            folder = argv[i + 1];
+        else if (option == "-T")
+            trial = argv[i + 1];
+        else if (option == "-S")
+            xsensReaderConfigFile = argv[i + 1];
+    }
+    try {
+        XsensDataReader reader(xsensReaderConfigFile);
+        if (!trial.empty()) reader.upd_trial_prefix() = trial;
+        if (!folder.empty()) reader.upd_data_folder() = folder;
+        DataAdapter::OutputTables tables = reader.readTrial();
+        // Write tables to sto files
+        // Accelerations
+        std::shared_ptr<AbstractDataTable> accelTable = tables.at(XsensDataReader::LinearAccelerations);
```

Comment:
Instead of the user/client of the `XsensDataReader` having to cast, it could cast abstract tables into their specific concrete types since it knows what the tables are. See `C3DFileAdapter::read()` for example.

Reply:
@aseth1 The signatures/names of custom accessors could look like this
`
     static const TimeSeriesTableQuaternion& getOrientationsTable(const DataAdapter::OutputTables& tables)

   static  const TimeSeriesTableVec3& getLinearAccelerationsTable(const DataAdapter::OutputTables& tables);
`
I opted to making the methods static and passing in tables as input over caching the tables as member variable  since  it relieves the class/object from memory management issues but I'd like to hear your input, and consistency considerations. Also caching and returning the map of Tables as done in C3DFileAdapter::read, allows the client to modify the Map which shouldn't be allowed AFAIK.

Reply:
Ideally individual table access for this reader and the `C3DFileAdapter` would be consistent. I don't think memory management is an issue using smart pointers internally, but I like the static methods and simplicity here. I am inclined to leave it as is as the static functions are easy to use from scripting as well.